src/bson.c

```diff
@@ -41,50 +41,201 @@
 #define PHONGO_BSON_STATE_ZCHILD(state) (((php_phongo_bson_state*) (state))->zchild)
 #endif
 
+#define PHONGO_FIELD_PATH_EXPANSION 8
+
 /* Forward declarations */
 static bool php_phongo_bson_visit_document(const bson_iter_t* iter ARG_UNUSED, const char* key, const bson_t* v_document, void* data);
 static bool php_phongo_bson_visit_array(const bson_iter_t* iter ARG_UNUSED, const char* key, const bson_t* v_document, void* data);
 
+/* Path builder */
+char* php_phongo_field_path_as_string(php_phongo_field_path* field_path)
+{
+	size_t length = 1; /* NULL character */
+	size_t i;
+	char*  path;
+	char*  ptr;
+
+	if (!field_path) {
+		return estrdup("");
+	}
+	if (!field_path->elements) {
+		return estrdup("");
+	}
+
+	for (i = 0; i <= field_path->current_level; i++) {
+		if (!field_path->elements[i]) {
+			continue;
+		}
+		length += (1 + strlen(field_path->elements[i]));
+	}
+
+	path = emalloc(length);
+	ptr  = path;
+
+	for (i = 0; i <= field_path->current_level; i++) {
+		if (!field_path->elements[i]) {
+			continue;
+		}
+		strcpy(ptr, field_path->elements[i]);
+		ptr += strlen(field_path->elements[i]);
+		ptr[0] = '.';
+		ptr++;
+	}
+	ptr[-1] = '\0';
+
+	return path;
+}
+
+php_phongo_field_path* php_phongo_field_path_alloc(void)
+{
+	php_phongo_field_path* tmp = ecalloc(1, sizeof(php_phongo_field_path));
+	tmp->ref_count             = 1;
+
+	return tmp;
+}
+
+void php_phongo_field_path_free(php_phongo_field_path* field_path)
+{
+	size_t i;
+
+	if (field_path->elements) {
+		for (i = 0; i < field_path->allocated_levels; i++) {
+			if (!field_path->elements[i]) {
+				continue;
+			}
+			field_path->elements[i] = NULL;
+		}
```

Comment:
I think we can do without this `for` loop, since we're going to free the entire `elements` array anyway. This may be a leftover from when you were actually freeing each element string.